include/envoy/http/header_map.h

```diff
@@ -348,6 +348,67 @@ class HeaderMap {
   virtual void addCopy(const LowerCaseString& key, const std::string& value) PURE;
 
   /**
+   * Set a reference header in the map. Both key and value MUST point to data that will live beyond
+   * the lifetime of any request/response using the string (since a codec may optimize for zero
+   * copy). Nothing will be copied.
+   *
+   * Calling setReference multiple times for the same header will result in only the last header
+   * being present in the HeaderMap.
+   *
+   * @param key specifies the name of the header to set; it WILL NOT be copied.
+   * @param value specifies the value of the header to set; it WILL NOT be copied.
+   */
+  virtual void setReference(const LowerCaseString& key, const std::string& value) PURE;
+
+  /**
+   * Set a header with a reference key in the map. The key MUST point to data that will live beyond
+   * the lifetime of any request/response using the string (since a codec may optimize for zero
+   * copy). The value will be copied.
+   *
+   * Calling setReferenceKey multiple times for the same header will result in only the last header
+   * being present in the HeaderMap.
+   *
+   * @param key specifies the name of the header to set; it WILL NOT be copied.
+   * @param value specifies the value of the header to set; it WILL be copied.
+   */
+  virtual void setReferenceKey(const LowerCaseString& key, uint64_t value) PURE;
+
+  /**
+   * Set a header with a reference key in the map. The key MUST point to point to data that will
+   * live beyond the lifetime of any request/response using the string (since a codec may optimize
+   * for zero copy). The value will be copied.
+   *
+   * Calling setReferenceKey multiple times for the same header will result in only the last header
+   * being present in the HeaderMap.
+   *
+   * @param key specifies the name of the header to set; it WILL NOT be copied.
+   * @param value specifies the value of the header to set; it WILL be copied.
+   */
+  virtual void setReferenceKey(const LowerCaseString& key, const std::string& value) PURE;
+
+  /**
+   * Set a header by copying both the header key and the value.
+   *
+   * Calling setCopy multiple times for the same header will result in only the last header being
+   * present in the HeaderMap.
+   *
+   * @param key specifies the name of the header to set; it WILL be copied.
+   * @param value specifies the value of the header to set; it WILL be copied.
+   */
+  virtual void setCopy(const LowerCaseString& key, uint64_t value) PURE;
```

Comment:
Are the two setCopy() calls added used anywhere?

Reply:
The string version is used in TestHeaderMapImpl to provide a convenience function for tests.

Reply:
Couldn't the test version just do remove followed by add? (I guess now that you removed update, couldn't it all just work that way)?

Reply:
Sure.

Reply:
It seems to me that having explicit method that do this work allows for future changes to optimize the set path, even if it's just remove/add under the covers now. But if you want, I can yank out all the set methods and move the remove calls into the router.

Reply:
You already wrote the optimized code but then you deleted it. 😉 I don't care either way. I mostly want to make sure that if we have functions they are actually used. If you want to keep the ones that are called but for example remove the setCopy stuff that isn't used other than test (and implement that in terms of remove/add) that's fine. LMK.

Reply:
Ok. I'll remove the unused ones.